### ISSUE TITLE
[Feature] Ask for new feature `train_evaluator`

### ISSUE THREAD

**Issue by JunweiZheng93** [open-mmlab/mmengine#1058]
### What is the feature?

We want to calculate metrics both in the training and validation process to check if the model is overfitted. We have `val_evaluator` and `test_evaluator` for metrics calculation in the validation and test process so far, but no `train_evaluator` during training. A workaround for me to calculate metrics during training is to do the calculation in `loss` mode and add the metrics to the loss_dict outputted by the `loss` mode. However, it will be great if `train_evaluator` can be implemented.

**Comment by HAOCHENYE** [comment]
If we can calculate the loss of validation set, will this be enough to check whether the model is overfitted? The overhead of calculating the validation loss is much smaller than calculating the training metric in most cases after all.

**Comment by JunweiZheng93** [comment]
HI @HAOCHENYE , thx for your quick reply.
1. Loss of validation set enough to check overfitting? I would say it is sufficient enough, but not perfect enough, since the metrics are always the most direct evaluations of the models. Losses are somehow in direct (or inverse) proportion to the metrics and the values of the losses are usually very small. It is also very hard to ensure the overfitting depending only on the losses.
2. Too much overhead? I totally agree with you. If the users prefer an `IterBasedTrainLoop`, that will be too expensive. In my case, I want to calculate the metrics on the training set after every training epoch, since I am a huge fan of `EpochBasedTrainLoop`. Therefore, the overhead is acceptable.

Thx for your time.

**Comment by HAOCHENYE** [comment]
Currently, this seems to be a relatively rare requirement, so we do not plan to implement it in the short term. You can customize your loop by overloading the `EpochBasedLoop`. Of course, if this requirement becomes a trend someday, we will support it as soon as possible. I will temporarily close this issue, but feel free to reopen it at any time 😄 .

**Comment by stwerner97** [comment]
Just wanted to add that I also would be interested in monitoring metrics on the training dataset. 😄
